docs/designs/mnnvl-design.md

```diff
@@ -0,0 +1,245 @@
+
+## Background
+
+### What is MNNVL
+
+**MNNVL (Multi-Node NVLink)** is an NVIDIA technology that extends NVLink-class, high-bandwidth GPU-to-GPU communication across **multiple physical nodes**, rather than being limited to GPUs within a single server. It uses specialized hardware and software mechanisms (such as IMEX for secure memory export/import) to allow GPUs on different nodes to access each other’s memory with much lower latency and higher bandwidth than traditional network-based approaches like TCP or even standard RDMA. In Kubernetes, MNNVL is exposed through NVIDIA’s DRA driver so distributed workloads (for example, large-scale training or tightly coupled inference) can treat GPUs across nodes as part of a single, high-performance compute fabric while preserving isolation and security between workloads.
+
+### Using MNNVL in a K8S cluster
+
+To use **MNNVL (Multi-Node NVLink)** in Kubernetes with NVIDIA’s DRA driver, you start by creating a **ComputeDomain (CD)**. The ComputeDomain represents a logical GPU fabric spanning multiple nodes and defines how inter-node NVLink/IMEX channels are allocated. In its `spec.channel`, the ComputeDomain references a **ResourceClaimTemplate**; this template describes the DRA resource class (MNNVL) that will be used to provision the interconnect. When the ComputeDomain is created, the NVIDIA DRA controller prepares the underlying GPU fabric and ensures that nodes participating in the domain can securely communicate over MNNVL.
+
+Next, pods that want to use MNNVL simply **reference the ResourceClaimTemplate** in their pod spec. Each pod declares a `resourceClaims` entry pointing to the template name, and the container lists that claim under `resources.claims`. Kubernetes then automatically creates a **ResourceClaim per pod** from the template. These per-pod claims are handed to the NVIDIA DRA driver, which allocates the necessary MNNVL/IMEX channels for each pod within the ComputeDomain.
+
+As a result, multiple pods—possibly scheduled on different nodes—are joined into the same ComputeDomain and can communicate using **multi-node NVLink semantics** without manually wiring GPUs or fabric resources. The ComputeDomain handles reachability and isolation, the ResourceClaimTemplate enables automatic per-pod allocation, and the pod spec remains simple: create a CD once, then reference its template in every pod that needs MNNVL.
+
+# Goals
+
+The following key goals, derived from the requirements document, guide the MNNVL design:
+* **Cluster Support:** The feature will only support homogeneous clusters.
+* **Global Configuration:** A global configuration for the feature is required. If the cluster does not support MNNVL, the operator must exit.
+* **Opt-Out and Workload Impact:**
```

Comment:
These are two different things and I would recommended to cover these in 2 different points.
`Opt-Out` - If MNNVL feature is enabled at the operator level then workload operators should be given a provision to opt out of automatic creation of DRA + ComputeDomain resources for specific workloads.